File: app/models/survey_mapping.rb
class SurveyMapping < ActiveRecord::Base
  belongs_to :survey
  belongs_to :story
  
  validates_presence_of :survey_id, :priority
  validates_numericality_of :priority

  # attr_accessible :survey_id, :story_id, :priority

  # Override as_json to exclude private attributes.
  def serializable_hash(options = {})
    if !options[:except]
      options[:except] = []
    end
    options[:except] << :id
    options[:except] << :survey_id
    if !options[:methods]
      options[:methods] = [:name, :description, :normalized_priority]
    end
    super(options)
  end
  
  def name
    story.name
  end
  
  def description
    story.description ? story.description : ''
  end
  
  def normalized_priority
    if story.priority && story.status_code < Story.Done
      ((((survey.survey_mappings.collect {|mapping| mapping.story}).select{|story| story.status_code<Story.Done}).collect{|story| story.priority}).sort).index(story.priority) + 1
    else
      nil
    end
  end
end

File: src/app/main/models/survey-mapping.ts
export class SurveyMapping {
  story_id: number;
  priority: number;
  name: string;
  description: string;
  normalized_priority: number;

  constructor(values: any) {
    this.story_id = values.story_id;
    this.priority = values.priority;
    this.name = values.name;
    this.description = values.description;
    this.normalized_priority = values.normalized_priority;
  }

  get descriptionFirstLine(): string {
    if (this.description) {
      let index: number = this.description.indexOf('\r');
      if (index > -1) {
        return this.description.substring(0, index);
      } else {
        return this.description;
      }
    } else {
      return this.description;
    }
  }
}
